Mathlib/Tactic/Linter/Header.lean

```diff
@@ -0,0 +1,281 @@
+* the `String` factor is the linter message.
+
+The linter checks that
+* the first and last line of the copyright are a `("/-", "-/")` pair, each on its own line;
+* the first line is `begins with `Copyright (c) 20` and ends with `. All rights reserved.`;
+* the second line is `Released under Apache 2.0 license as described in the file LICENSE.`;
+* the remainder of the string begins with `Authors: `, does not end with `.` and
+  contains no ` and ` nor a double space, except possibly after a line break.
+-/
+def copyrightHeaderChecks (copyright : String) : Array (Syntax × String) := Id.run do
+  -- first, we merge lines ending in `,`: two spaces after the line-break are ok,
+  -- but so is only one or none.  We take care of *not* adding more consecutive spaces, though.
+  let preprocessCopyright := (copyright.replace ",\n  " ", ").replace ",\n" ","
+  -- filter out everything after the first isolated `-/`
+  let pieces := preprocessCopyright.splitOn "\n-/"
+  let copyright := (pieces.getD 0 "") ++ "\n-/"
+  let stdTxt (s : String) :=
+    s!"Malformed or missing copyright header: `{s}` should be alone on its own line."
+  let mut msgs := #[]
+  if (pieces.getD 1 "\n").take 1 != "\n" then
+    msgs := msgs.push (toSyntax copyright "-/", s!"{stdTxt "-/"}")
+  let lines := copyright.splitOn "\n"
+  let closeComment := lines.getLastD ""
+  match lines with
+  | openComment :: copyrightAuthor :: license :: authorsLines =>
+    -- The header should start and end with blank comments.
+    match openComment, closeComment with
+    | "/-", "-/" => msgs := msgs
+    | "/-", _    =>
+      msgs := msgs.push (toSyntax copyright closeComment, s!"{stdTxt "-/"}")
+    | _, _       =>
+      msgs := msgs.push (toSyntax copyright openComment, s!"{stdTxt ("/".push '-')}")
+    -- validate copyright author
+    let copStart := "Copyright (c) 20"
+    let copStop := ". All rights reserved."
+    if !copyrightAuthor.startsWith copStart then
+      msgs := msgs.push
+        (toSyntax copyright (copyrightAuthor.take copStart.length),
+         s!"Copyright line should start with 'Copyright (c) YYYY'")
+    if !copyrightAuthor.endsWith copStop then
+      msgs := msgs.push
+        (toSyntax copyright (copyrightAuthor.takeRight copStop.length),
+         s!"Copyright line should end with '. All rights reserved.'")
+    -- validate authors
+    let authorsLine := "\n".intercalate authorsLines.dropLast
+    let authorsStart := (("\n".intercalate [openComment, copyrightAuthor, license, ""])).endPos
+    for corr in authorsLineChecks authorsLine authorsStart do
+      msgs := msgs.push corr
+    let expectedLicense := "Released under Apache 2.0 license as described in the file LICENSE."
+    if license != expectedLicense then
+      msgs := msgs.push (toSyntax copyright license,
+        s!"Second copyright line should be \"{expectedLicense}\"")
+  | _ =>
+    msgs := msgs.push (toSyntax copyright "-/", s!"Copyright too short!")
```

Comment:
I tried adding a test which would hit this path, and failed: can you verify if there's a bug somewhere?

The following does *not* error, which I find surprising.
```
#guard_msgs in
#check_copyright
"/-
Copyright (c) 2024 Damiano Testa. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Name LastName
stuff
-/
"
```

The following is also surprising: it complains about the authors line...
#guard_msgs in
#check_copyright
"/-
Copyright (c) 2024 Damiano Testa. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
-/
"

Reply:
I added a test that emits `Copyright too short!` as a message.

For the rest, I am not sure how much more we should give hints: really, the "fix" to all these issues is to use the template... and not mess up!

Reply:
That is true.
Re-thinking this, I *do* care about one edge case: about hidden doc comments being snuck into the copyright header. I made the linter detect those, and then finding a missing authors line was easy, so I added this. I pushed these as new commits.